src/sensor_hx71x.c

```diff
@@ -0,0 +1,267 @@
+{
+    while (!hx71x_check_elapsed(start, hx71x_get_time(), ticks))
+        ;
+}
+
+static inline void
+hx71x_delay(hx71x_time_t start, hx71x_time_t ticks)
+{
+    while (!hx71x_check_elapsed(start, hx71x_get_time(), ticks))
+        irq_poll();
+}
+
+#endif
+
+/****************************************************************
+ * HX711 and HX717 Sensor Support
+ ****************************************************************/
+// both HX717 and HX711 have 200ns min pulse time for clock pin on/off
+#define MIN_PULSE_TIME nsecs_to_ticks(200)
+#define MAX_READ_TIME timer_from_us(50)
+
+// Event handler that wakes wake_hx71x() periodically
+static uint_fast8_t
+hx71x_event(struct timer *timer)
+{
+    struct hx71x_adc *hx71x = container_of(timer, struct hx71x_adc, timer);
+    hx71x->pending_flag = 1;
+    sched_wake_task(&wake_hx71x);
+    return SF_DONE;
+}
+
+// Helper code to reschedule the hx71x_event() timer
+static void
+hx71x_reschedule_timer(struct hx71x_adc *hx71x)
+{
+    irq_disable();
+    hx71x->pending_flag = 1;
+    hx71x->timer.waketime = timer_read_time() + hx71x->rest_ticks;
+    sched_add_timer(&hx71x->timer);
+    irq_enable();
+}
+
+int8_t
+hx71x_is_data_ready(struct hx71x_adc *hx71x) {
+    // if the dout pin is high the sample is not ready
+    return !gpio_in_read(hx71x->dout);
+}
+
+// Add a measurement to the buffer
+static void
+add_sample(struct hx71x_adc *hx71x, uint_fast32_t counts)
+{
+    hx71x->sb.data[hx71x->sb.data_count] = counts;
+    hx71x->sb.data[hx71x->sb.data_count + 1] = counts >> 8;
+    hx71x->sb.data[hx71x->sb.data_count + 2] = counts >> 16;
+    hx71x->sb.data[hx71x->sb.data_count + 3] = counts >> 24;
+    hx71x->sb.data_count += BYTES_PER_SAMPLE;
+}
+
+static void
+flush_samples(struct hx71x_adc *hx71x, uint8_t oid)
+{
+    if (hx71x->sb.data_count + BYTES_PER_SAMPLE > ARRAY_SIZE(hx71x->sb.data))
+        sensor_bulk_report(&hx71x->sb, oid);
+}
+
+// Pulse all clock pins to move to the next bit
+inline static void
+hx71x_pulse_clocks(struct hx71x_adc *hx71x) {
+    irq_disable();
+    gpio_out_write(hx71x->sclk, 1);
+    hx71x_delay_no_irq(hx71x_get_time(), MIN_PULSE_TIME);
+    gpio_out_write(hx71x->sclk, 0);
+    hx71x_delay_no_irq(hx71x_get_time(), MIN_PULSE_TIME);
+    irq_enable();
+}
+
+// hx71x ADC query
+void
+hx71x_read_adc(struct hx71x_adc *hx71x, uint8_t oid)
+{
+    if (!hx71x_is_data_ready(hx71x)) {
+        hx71x_reschedule_timer(hx71x);
+        return;
+    }
+
+    // data is ready
+    int32_t counts = 0;
+    hx71x_time_t start_time = timer_read_time();
+    for (uint_fast8_t sample_idx = 0; sample_idx < 24; sample_idx++) {
+        hx71x_pulse_clocks(hx71x);
+        // read 2's compliment int bits
+        counts = (counts << 1) | gpio_in_read(hx71x->dout);
+    }
+
+    // bit bang 1 to 4 more bits to configure gain & channel for the next sample
+    for (uint8_t gain_idx = 0; gain_idx < hx71x->gain_channel; gain_idx++) {
+        hx71x_pulse_clocks(hx71x);
+    }
```

Comment:
As a minor note, this code could be optimized a bit.  Something to look at after merging though.